tools/tplgtool.py

```diff
@@ -607,6 +607,31 @@ def _to_fmt_string(fmt):
             fmts.append("FLOAT")
         return fmts
 
+
+    # transform number denoted pipeline format to string
+    @staticmethod
+    def _to_rate_string(rate):
+        rates = []
+        # Fred: should we check highest freqency first to make decending order?
+        # Fred: is this all the rates supported?
+        if rate & (1 << 1) != 0:
+            rates.append("8000")
+        if rate & (1 << 3) != 0:
+            rates.append("16000")
+        if rate & (1 << 4) != 0:
+            rates.append("22050")
+        if rate & (1 << 5) != 0:
+            rates.append("32000")
+        if rate & (1 << 6) != 0:
+            rates.append("44100")
+        if rate & (1 << 7) != 0:
+            rates.append("48000")
+        if rate & (1 << 8) != 0:
+            rates.append("64000")
+        if rate & (1 << 9) != 0:
+            rates.append("96000")
```

Comment:
This is a lot of repetition; I think you don't need to "unroll" that loop:

```python
  bit_rates_dict = { 1 : 8000, 3 : 16000, ...}

  for bit in rates_dict:
     if rate & (1 << bit) != 0:
        rates.append(bit_rates_dict[bit])

```
Or even shorter

```python
  rates = [ bit_rates_dict[bit] for bit in bit_rates_dict
                   if rate & (1 << bit) != 0 ]
```

Reply:
This is really good comment, but I can't get it to work. I put a TODO above. There are a couple of places where the same thing is applicable. Can I make this improvement as a separate PR?

Reply:
What is the error?

Reply:
I tried to test this to help but the new functions are never called. Did you forget a part?

Reply:
> I tried to test this to help but the new functions are never called. Did you forget a part?

OK I figured it out: the "interactive" output of the parser is incomplete.


I just tested this below together with https://github.com/thesofproject/sof/pull/9017/files and it works perfectly fine:

```python
    @staticmethod
    def _to_rate_string(rate):
        # Note: intentionally put 48000 first to make first in the list
        bit_rates_dict = { 7 : 48000, 6 : 44100, }

        return [ str(bit_rates_dict[bit]) for
                 bit in bit_rates_dict if rate & (1 << bit) != 0 ]
```

I forgot the `str()` earlier, is that why it was failing for you?

Reply:
wow, it works nicely. I just updated again. Let me check the test results before remove draft